Add "How to write a how-to" guide for developers (#181)
napari/napari.github.io: developers/contributing_how_tos.md

# How to contribute a napari How-To

> This should be a sentence (or brief paragraph) about what this how-to will achieve and how it helps address a common problem/task/need.

Reviewer: what the how-to will achieve doesn't feel super clear to me, is it about what the document itself will demonstrate? Maybe demonstrate is a clearer word (or I'm just slow 🙂)

Reviewer 2: This can be stated as what the reader will be able to do after reading through the document. I like to focus these statements in terms of the reader and not the document. So, I wouldn't say "what the how-to will achieve" but "what the reader will achieve after reading this how-to" (maybe too subtle but I feel it helps to pinpoint the goal of each document).

## Prerequisites

- this template
- a list of what the user needs to prepare to be ready to start this how-to guide 
- links to any existing napari.org tutorials or how-to guides that can help the user fulfill these prerequisites
- the level of python/napari knowledge required to follow this how-to
    - try to be specific about what skills are needed e.g. 

>     'connecting callbacks to layer events' or 'using matplotlib to produce plots'

Reviewer: super important, love this!

- plugins that should be installed
- python packages that should be installed (don't list napari dependencies)

***

## How-to steps

> Headings in this section are yours to decide on. You should split your how-to guide into easy to follow, sequential steps to achieve the stated goal. Remember that your how-to may be used as a reference, so your headings should be descriptive and allow users to get a quick overview of the document, and which steps are relevant to them.

Reviewer: I like the idea of having an `outline` section before the actual steps in the how-to document. This section would quickly lists the steps, providing a super short summary of the steps to be taken and, if appropriate, a little more detail than the title of the section.

Reviewer 2: I think this will be automatically generated by the headings in the right-side toolbar, so not sure we need to include this again in the document itself.

Reviewer: @melissawm you're right - I was initially thinking that for a longer form how-to this could be useful for providing a bit more detail but actually how-tos should probably remain very limited in scope

Author: @alisterburt I think that sort of outline/longer form intro would be perfect for a tutorial, but would expect how-tos to have limited enough scope for a brief paragraph to be sufficient (if a sentence isn't quite).

## Include pictures

> You should include pictures/videos, particularly when describing interactions with the viewer. If an action can be performed both through the viewer and through code, you should include both options.

Reviewer: <3

Reviewer 2: (should we include a mention of accessibility - "try to use descriptive and complete alt-text when including images"?)

## How-to vs. tutorial

> A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how. A how-to does not explain concepts or engage in discussion, and should prioritise practical usability. In comparison, a tutorial is a broader exploration, or guided tour of what is available. It might include detailed explanations and discussion, and is usually aimed at users who don't yet know what can be done.

Reviewer 3: I think it would be clearer to focus on what to do, or what makes a good how-to document.

It becomes more confusing for readers if there's a lot of:
* Sentences saying "don't do X or Y" (don't think of unicorn! ...now that's all you can think about, right?)
* Sentences trying to clarify the difference between a how-to and a tutorial

... you know, I think what I'm saying here is that you are inserting too much explanation content into your own how-to guide

Author: > ... you know, I think what I'm saying here is that you are inserting too much explanation content into your own how-to guide

Yes, I too had this feeling. I will reword to focus more on the how-to rather than how-not-to, and we can maybe work on disambiguation of tutorials and how-to's elsewhere

Reviewer: +1, great suggestion @GenevieveBuckley - I think this section reads very nicely already if you just remove the 'don't do this' parts:

edit: maybe this should be moved to the top?
```suggestion

## How-to vs. tutorial

A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how.
```

Reviewer 2: I think this doesn't need to be removed, actually. I quite like the idea of making sure people understand this from the beginning. I'd suggest:

```suggestion
A how-to is a step by step guide designed to answer a specific "How do I..." question. This means the user already knows **what** they want to do, just not how. A how-to does not explain concepts or engage in discussion, and should prioritise practical usability.
```

This sentence:
```
In comparison, a tutorial is a broader exploration, or guided tour of what is available. 
It might include detailed explanations and discussion, and is usually aimed at users
who don't yet know what can be done.
```
could go into the tutorial exemplar/guide.

> 

Author: We may also want some style guide suggestions here? I'm not sure if we want to be that prescriptive but for example the [Google documentation style guide](https://developers.google.com/style/) which they use over at [numpy](https://numpy.org/numpy-tutorials/content/tutorial-style-guide.html#make-use-of-the-google-doc-style-guide) would be an option

Reviewer: good idea - this is good to highlight!

Reviewer 2: +1 for the google documentation style guide - I think it is very direct and informative without being over restrictive. They also have amazing [technical writing guides](https://developers.google.com/tech-writing) which I highly recommend.

## Where does this how-to go?

> Once you've written your how-to, you should place this document under `getting_started/how_tos/` and open a PR in the [napari.github.io](https://github.com/napari/napari.github.io) repository.

Author: We will probably need to update this to expand a bit more on where in the repository this document could go as I believe there are two sections for how-tos @melissawm?

Reviewer 2: Actually, I didn't anticipate a separate how-to section under the "Getting started" section. But I'm happy to discuss this if you feel it's needed

Author: No, I think I just got confused because the getting started section contains some how-tos. But I can see that this section is unlikely to be contributed to, so I guess we direct them to place their new document in a new folder we don't have yet called 'how-to's?

***

## Further Reading

> If you felt robbed of your chance to provide explanations, this is your opportunity to link out to useful references or discussions. You should also link to any relevant napari.org how-tos, tutorials or examples that can help the user learn more about the topic.

Reviewer 2: +1, this is the way to go - delegate longer, deeper explanations elsewhere!

- The [Diátaxis Framework](https://diataxis.fr/) is the basis for how we are organising our documentation

> - This [Divio document](https://documentation.divio.com/how-to-guides/) explains how-to guides in more detail

Author: Once this is closer to being finalised we should also link to a good "how-to" example of our own as an exemplar.

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.13.0
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

# How to contribute a napari How-To

> This should be a sentence (or brief paragraph) about what the reader will achieve after following this

Reviewer 3: I think this phrasing is a bit too meta, like it's conflating this guide with a template in itself.
It reads more like "we meant to put content here at this placeholder but forgot", rather than the message you want to convey.
